smartsim/ray/raycluster.py

```diff
@@ -0,0 +1,378 @@
+        """Get the ray head node host address from the log file produced
+        by the head process.
+
+        :return: address of the head host
+        :rtype: str
+        """
+        
+#         if self._launcher == 'local':
+#             self.head_model.address = '127.0.0.1'
+#             return
+        
+        # We can rely on the file name, because we set it when we create
+        # the head model
+        head_log = os.path.join(self.head_model.path, "head.out")
+        
+        max_attempts = 60
+        attempts = 0
+        while not os.path.isfile(head_log):
+            time.sleep(1)
+            attempts += 1
+            if attempts==max_attempts:
+                raise RuntimeError("Could not find Ray cluster head address.")
+                
+        attempts = 0
+        head_ip = None
+        while head_ip is None:
+            time.sleep(1)
+            with open(head_log) as fp:
+                line = fp.readline()
+                while line:
+                    plain_line = re.sub('\033\\[([0-9]+)(;[0-9]+)*m', '', line) 
+                    if "Local node IP:" in plain_line:
+                        matches=re.search(r'(?<=Local node IP: ).*', plain_line)
+                        head_ip = matches.group()
+                    line = fp.readline()       
+            attempts += 1
+            if attempts==max_attempts:
+                raise RuntimeError("Could not find Ray cluster head address.")
+                
+        self.head_model.address = head_ip
+        
+        
+class RayHead(Model):
+    """Ray head node model.
+    
+    :param name: The name of the entity.
+    :type name: str
+    :param params: model parameters for writing into configuration files.
+    :type params: dict
+    :param path: path to output, error, and configuration files
+    :type path: str
+    :param ray_password: Password used to connect to Redis by Ray
+    :type ray_password: str
+    :param ray_port: Port at which the head node will be running
+    :type ray_port: int
+    :param ray_num_cpus: Number of CPUs used by Ray
+    :type ray_num_cpus: int
+    :param run_args: Arguments to pass to launcher to specify details such as partition, time, and so on
+    :type run_args: dict[str,str]
+    :param launcher: Name of launcher to use for starting the cluster
+    :type launcher: str
+    :param alloc: ID of allocation to run on, only used if launcher is Slurm and allocation is
+                  obtained with ``ray.slurm.get_allocation``
+    :type alloc: int
+    :param batch: Whether the head node should be launched through a batch file
+    :type batch: bool
+    :param time: The walltime the head node will be running for
+    :type time: str
+    """
+    def __init__(self, 
+                 name, 
+                 params, 
+                 path, 
+                 ray_password, 
+                 ray_port=6780, 
+                 run_args={}, 
+                 ray_num_cpus=12, 
```

Comment:
Im going to argue the default should be lower than this. esp for local launching. maybe 4?